rollup/pipeline/pipeline.go

```diff
@@ -199,25 +205,47 @@ type BlockCandidate struct {
 	CoalescedLogs []*types.Log
 }
 
+// sendCancellable tries to send msg to resCh but allows send operation to be cancelled
+// by closing cancelCh. Returns true if cancelled.
+func sendCancellable[T any, C comparable](resCh chan T, msg T, cancelCh <-chan C) bool {
+	var zeroC C
+
+	select {
+	case resCh <- msg:
+		return false
+	case cancelSignal := <-cancelCh:
+		if cancelSignal != zeroC {
+			panic("shouldn't have happened")
+		}
+		return true
+	}
+}
+
 func (p *Pipeline) traceAndApplyStage(txsIn <-chan *types.Transaction) (<-chan error, <-chan *BlockCandidate, error) {
 	p.state.StartPrefetcher("miner")
-	newCandidateCh := make(chan *BlockCandidate)
+	downstreamCh := make(chan *BlockCandidate, p.downstreamChCapacity())
```

Comment:
Are there any concerns about storing `N` traces in memory here?

Reply:
How big can a trace get?

Reply:
a typical big trace is hundreds of MB

Reply:
> a typical big trace is hundreds of MB

that is per block.

> Are there any concerns about storing N traces in memory here?

traces here are per txn, so it is no diffferent than storing the trace of the entire block in memory for CCC. Like follower nodes do.

Reply:
> a typical big trace is hundreds of MB

that must be per block if I am not mistkaen.